### DISCUSSION TITLE
Ready-to-use in-browser terminal emulation? (history, etc)

### DISCUSSION THREAD

**Discussion by iliakan** [sourcelair/xterm.js#4404] [General]
Is there a ready-to-use in-browser terminal emulation, browsed on xterm.js?

I mean, with history support, standard key combinations, some other functionality maybe?

Totally in-browser.

**Comment by Tyriar** [comment]
A couple of WIP projects you might find useful:

- I have a project called "shell engine" with the goal being what you're after https://github.com/microsoft/shell-engine-js with a particular focus on use with the `Pseudoterminal` extension API in VS Code. It's a little rough around the edges and not yet published to npm, but you could add it as a git submodule to use.
- Pty emulation from @jerch which aims to act closer to an actual terminal/pty https://github.com/jerch/browser-fakepty

**Reply by iliakan** [reply]
Thank you, the second project is very outdated and given up for sure.

The shell-engine-js looks better, but I couldn't build/run it.

Anything else maybe?

**Reply by Tyriar** [reply]
There are some others if you search "xterm local echo", I haven't tried them out though.
